src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,424 @@
+# in this way we can easily later decide on the codepath using type signatures
+
+"""
+    ColRename
+
+A singleton type indicating that column renaming operation was requested in `select`.
+"""
+struct ColRename end
+
+normalize_selection(idx::AbstractIndex, sel) = idx[sel]
+
+function normalize_selection(idx::AbstractIndex, sel::ColumnIndex)
+    loc = idx[sel]
+    return loc => ColRename() => _names[loc]
+end
+
+normalize_selection(idx::AbstractIndex, sel::Pair{<:Any,<:Pair{<:Base.Callable,Symbol}}) =
+    idx[first(sel)] => last(sel)
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    newcol = Symbol(_names(idx)[c], "_", funname(fun))
+    return c => fun => newcol
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    if length(c) > 3
+        newcol = Symbol(join(_names(idx)[c[1:2]], "_"), "_etc_", funname(fun))
+    else
+        newcol = Symbol(join(_names(idx)[c], "_"), "_", funname(fun))
+    end
+    return c => fun => newcol
+end
+
+function select_transform!(nc::Union{Pair{Int, Pair{ColRename, Symbol}},
+                                     Pair{<:Union{Int, AbstractVector{Int}},
+                                          <:Pair{<:Base.Callable, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Symbol, Any}, copycols::Bool)
+    col_idx = first(nc)
+    transform_spec = last(nc)
+    newname = last(transform_spec)
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        newdf[!, newname] = first(transform_spec).(df[!, col_idx])
+    elseif nc isa Pair{<:AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+        if length(col_idx) == 0
+            newdf[!, newname] = [first(transform_spec)() for _ in axes(df, 1)]
+        else
+            rowiterator = Tables.rows(Tables.columntable(df[!, col_idx]))
+            newdf[!, newname] = map(first(transform_spec),
+                                    Tables.namedtupleiterator(eltype(rowiterator), rowiterator))
+        end
+    else
+        throw(ErrorException("code should never reach this branch"))
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing
+provided that the columns requested in each of them are unique and present in `df`.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Column renaming and transformations are supported.
```

Comment:
Maybe this?
```
Columns can be renamed using the `old_column => new_column_name` syntax,
and transformed using the `old_column => fun => new_column_name` syntax.
`new_column_name` must be a `Symbol`, and `fun` a function or a type.
If `old_column` is a `Symbol` or an integer then `fun` is applied to each element
(row) of `old_column`.
Otherwise `old_column` can be any column indexing syntax, but in this case `fun`
will be passed a `NamedTuple` representing each row, holding only
the columns specified by `old_column`.

Column transformation can also be specified using the short `old_column => fun` form.
In this case, `new_column_name` is automatically generated as `$(old_column)_$(fun)`
(only the two first column names are used for multiple input columns).
```

Reply:
Thank you for an excellent write-up. I have just updated it with a more precise description of automatic column generation rules.